File: programmers/programmers_81303.py
def solution(n, k, cmd):
  
    answer = ''
    stack = []

    for c in cmd : 

        if c[0] == 'U' :
            cnt = int(c[2:])
            k -= cnt

        elif c[0] == 'D' :
            cnt = int(c[2:])
            k += cnt

        elif c[0] == 'C' : 
            stack.append(k)
            n -= 1
            if k == n : k -= 1

        elif c[0] == 'Z' :
            ele = stack[-1]
            del stack[-1]
            if k >= ele : k += 1
            n += 1

    for i in range(n):
        answer += 'O'

    while stack : 
        s = stack[-1]
        del stack[-1]
        answer = answer[:s] + 'X' + answer[s:]
    
    return answer

File: programmers/programmers_81303.kt
class Solution_81303 {

    var list = mutableListOf<String>()
    var table = mutableListOf<String>()
    var del = mutableListOf<Pair<String, Int>>()

    fun solution(n: Int, k: Int, cmd: Array<String>): String {
        var curr = k
        (0..n - 1).forEach { it ->
            table.add(it.toString())
            list.add("X")
        }

        cmd.forEach { order ->
            when(order[0]) {
                'U' -> {
                    curr -= order.substring(2).toInt()
                }
                'D' -> {
                    curr += order.substring(2).toInt()
                }
                'C' -> {
                    del.add(Pair(table[curr], curr))
                    table.removeAt(curr)
                    if (curr == table.size) curr -= 1
                }
                'Z' -> {
                    val data = del.last()
                    del.remove(data)
                    if (data.second <= curr) curr += 1
                    table.add(data.second, data.first)
                }
            }
        }

        table.forEach { 
            list[it.toInt()] = "O"
        }
        val answer = list.joinToString("") 

        return answer
    }
}